src/impl/pooling_direct.jl

```diff
@@ -41,69 +41,80 @@ for name in (:max, :mean)
 
         # Start with the central region
         w_region, h_region, d_region = central_region
-        @inbounds for batch_idx in 1:size(x)[end],
-            c in 1:out_c,
-            d in d_region,
-            h in h_region,
-            w in w_region
-          
-            # Initialize `m` to `0.0`, or `typemin(T)` or whatever.
+
+        @inbounds for batch_idx in 1:size(x, 5), c in 1:out_c
+            for d in d_region
+            pd = project(d, stride_d, pad_d_lo)
+            for h in h_region
+            ph = project(h, stride_h, pad_h_lo)
+            for w in w_region
+            pw = project(w, stride_w, pad_w_lo)
             m = m_init
 
             for kd in 1:kernel_d,
                 kh in 1:kernel_h,
                 kw in 1:kernel_w
 
-                input_kd = project(d, stride_d, pad_d_lo) + (kd - 1) * dil_d
-                input_kh = project(h, stride_h, pad_h_lo) + (kh - 1) * dil_h
-                input_kw = project(w, stride_w, pad_w_lo) + (kw - 1) * dil_w
+                input_kd = pd + (kd - 1) * dil_d
+                input_kh = ph + (kh - 1) * dil_h
+                input_kw = pw + (kw - 1) * dil_w
 
                 # This conditional will be optimized away at compile time
                 if $(name == :max)
-                    m = max(m, x[input_kw, input_kh, input_kd, c, batch_idx])
+                    xv = x[input_kw, input_kh, input_kd, c, batch_idx]
+                    if xv > m
+                        m = xv
+                    end
```

Comment:
Is this faster? If so, might `m = ifelse(xv > m, xv, m)` be even faster?

Reply:
My assumption was that if `xv < m`, then we can avoid assignment at all